src/common/operator/cast_operators.cpp

```diff
@@ -1579,11 +1608,45 @@ struct DecimalCastOperation {
 		return true;
 	}
 
-	template <class T>
+	template <class T, bool NEGATIVE>
+	static bool TruncateExcessiveDecimals(T &state) {
+		D_ASSERT(state.excessive_decimals);
+		bool round_up = false;
+		for (idx_t i = 0; i < state.excessive_decimals; i++) {
+			auto mod = state.result % 10;
+			round_up = NEGATIVE ? mod <= -5 : mod >= 5;
+			state.result /= 10.0;
+		}
+		//! Only round up when exponents are involved
+		if (state.exponent && round_up) {
+			if (NEGATIVE) {
+				state.result -= 1;
+			} else {
+				state.result += 1;
+			}
+		}
+		D_ASSERT(state.decimal_count > state.scale);
+		state.decimal_count = state.scale;
+		return true;
+	}
+
+	template <class T, bool NEGATIVE>
 	static bool Finalize(T &state) {
-		// if we have not gotten exactly "scale" decimals, we need to multiply the result
-		// e.g. if we have a string "1.0" that is cast to a DECIMAL(9,3), the value needs to be 1000
-		// but we have only gotten the value "10" so far, so we multiply by 1000
+		if (state.excessive_decimals) {
+			//! Encountered an exponent
+			if (!TruncateExcessiveDecimals<T, NEGATIVE>(state)) {
```

Comment:
if the 'else if' case is checked before this, then the two TruncateExcessiveDecimals calls can be unified